### PR TITLE
Promtool: Add support for compaction analysis

### PR DESCRIPTION
**Pull request by fpetkovski**
This commit extends the `promtool tsdb analyze` command to help 
troubleshoot high Prometheus disk usage. The command will now plot a distribution of 
how full chunks are relative to the maximum capacity of 120 samples per chunk.

Here is a sample output:
```
$ promtool tsdb analyze

...

Fullness: Amount of chunks
     10%: #############
     20%: ######
     30%: 
     40%: 
     50%: #####################################
     60%: ######
     70%: 
     80%: 
     90%: 
    100%: ###################################
```

### CODE REVIEW

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -142,6 +142,10 @@ func main() {
 	dumpMinTime := tsdbDumpCmd.Flag("min-time", "Minimum timestamp to dump.").Default(strconv.FormatInt(math.MinInt64, 10)).Int64()
 	dumpMaxTime := tsdbDumpCmd.Flag("max-time", "Maximum timestamp to dump.").Default(strconv.FormatInt(math.MaxInt64, 10)).Int64()
 
+	tsdbAnalyzeCompaction := tsdbCmd.Command("analyze-compaction", "Plot a distribution of how full TSDB chunks are, relative to the maximum capacity of 120 samples.")
+	analyzeCompactionDB := tsdbAnalyzeCompaction.Arg("db path", "Database path (default is "+defaultDBPath+").").Default(defaultDBPath).String()
+	analyzeCompactionBlockID := tsdbAnalyzeCompaction.Arg("block ID", "The TSDB block to analyze").Default("").String()
```
Comment (fpetkovski) [inline cmd/promtool/main.go:147]: We could also consider adding this analysis to the existing `tsdb analyze` command.

Comment (simonpasquier) [reply cmd/promtool/main.go:147]: IMHO it would be fit in the scope of `tsdb analyze` but lets see what others think :)

Comment (fpetkovski) [reply cmd/promtool/main.go:147]: @dgl do you have any thoughts on this one?

**Comment by bwplotka** [comment]
PTAL @codesome @cstyan , I will look in my free time too! 🙈

File: cmd/promtool/main.go (hunk 1 above)
Comment (dgl) [reply cmd/promtool/main.go:147]: Yeah, makes sense to include it to me.

Comment (bwplotka) [reply cmd/promtool/main.go:147]: +1 can we merge to tsdb analyze then? 🤗

**Review by bwplotka — APPROVED** [review]
Nice, LGTM when added to `tsdb analyze` (:

**Comment by fpetkovski** [comment]
@bwplotka @dgl the compaction analysis should now be merged into the `tsdb analyze` command.

File: cmd/promtool/tsdb.go [hunk 2]
```diff
@@ -561,6 +562,60 @@ func analyzeBlock(path, blockID string, limit int) error {
...
+			// Calculate the bucket for the chunk and increment it in the histogram
+			bucket := int(math.Ceil(float64(nBuckets)*chunkSize/maxSamplesPerChunk)) - 1
+			histogram[bucket]++
+			totalChunks++
+		}
+	}
+
+	fmt.Printf("\nCompaction analysis:\n")
+	fmt.Println("Fullness: Amount of chunks")
+	// Normalize absolute counts to percentages and print them out
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:609]:
```suggestion
	// Normalize absolute counts to percentages and print them out.
```

File: cmd/promtool/tsdb.go [hunk 3]
```diff
@@ -561,6 +562,60 @@ func analyzeBlock(path, blockID string, limit int) error {
...
+		}
+
+		for _, chk := range chks {
+			// Load the actual data of the chunk
+			chk, err := chunkr.Chunk(chk.Ref)
+			if err != nil {
+				return err
+			}
+			chunkSize := math.Min(float64(chk.NumSamples()), maxSamplesPerChunk)
+			// Calculate the bucket for the chunk and increment it in the histogram
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:600]:
```suggestion
			// Calculate the bucket for the chunk and increment it in the histogram.
```

File: cmd/promtool/tsdb.go [hunk 4]
```diff
@@ -561,6 +562,60 @@ func analyzeBlock(path, blockID string, limit int) error {
...
+	totalChunks := 0
+	for postingsr.Next() {
+		var lbsl = labels.Labels{}
+		var chks []chunks.Meta
+		if err := indexr.Series(postingsr.At(), &lbsl, &chks); err != nil {
+			return err
+		}
+
+		for _, chk := range chks {
+			// Load the actual data of the chunk
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:594]:
```suggestion
			// Load the actual data of the chunk.
```

File: cmd/promtool/tsdb.go [hunk 5]
```diff
@@ -561,6 +562,60 @@ func analyzeBlock(path, blockID string, limit int) error {
 	}
 	fmt.Printf("\nHighest cardinality metric names:\n")
 	printInfo(postingInfos)
+
+	return analyzeCompaction(block, ir)
+}
+
+func analyzeCompaction(block tsdb.BlockReader, indexr tsdb.IndexReader) (err error) {
+	postingsr, err := indexr.Postings("", "")
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:570]:
```suggestion
	postingsr, err := indexr.Postings(index.AllPostingsKey())
```

**Review by bwplotka — APPROVED** [review]
Nice! LGTM, just small tiny nits. 

Plus: Can we mention new functionality in `tsdb analysis` command help?

**Review by bwplotka — APPROVED** [review]

File: cmd/promtool/tsdb.go [hunk 6]
```diff
@@ -561,6 +563,60 @@ func analyzeBlock(path, blockID string, limit int) error {
...
+			chunkSize := math.Min(float64(chk.NumSamples()), maxSamplesPerChunk)
+			// Calculate the bucket for the chunk and increment it in the histogram.
+			bucket := int(math.Ceil(float64(nBuckets)*chunkSize/maxSamplesPerChunk)) - 1
+			histogram[bucket]++
+			totalChunks++
+		}
+	}
+
+	fmt.Printf("\nCompaction analysis:\n")
+	fmt.Println("Fullness: Amount of chunks")
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:609]:
```suggestion
	fmt.Println("Fullness: Amount of samples in chunks (100% is 120 samples)")
```

**Review by bwplotka — APPROVED** [review]
One more thing ;p

**Comment by bwplotka** [comment]
Thanks!

**Comment by roidelapluie** [comment]
This pull request seems a massive regression on the time spent by tsdb analysis. I think we should put this option behind a flag.

**Comment by fpetkovski** [comment]
Hi @roidelapluie, thanks for the review. What do you think about an `--extended` flag which would enable the compaction analysis?

**Pull request merged**